qiskit/dagcircuit/collect_blocks.py

```diff
@@ -0,0 +1,255 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Various ways to divide a DAG into blocks of nodes, to split blocks of nodes
+into smaller sub-blocks, and to consolidate blocks."""
+
+from qiskit.circuit import QuantumCircuit, CircuitInstruction
+from . import DAGOpNode, DAGCircuit, DAGDependency
+from .exceptions import DAGCircuitError
+
+
+class BlockCollector:
+    """This class implements various strategies of dividing a DAG (direct acyclic graph)
+    into blocks of nodes that satisfy certain criteria. It works both with the
+    :class:`~qiskit.dagcircuit.DAGCircuit` and
+    :class:`~qiskit.dagcircuit.DAGDependency` representations of a DAG, where
+    DagDependency takes into account commutativity between nodes.
+
+    Collecting nodes from DAGDependency generally leads to more optimal results, but is
+    slower, as it requires to construct a DAGDependency beforehand. Thus, DAGCircuit should
+    be used with lower transpiler settings, and DAGDependency should be used with higher
+    transpiler settings.
+
+    In general, there are multiple ways to collect maximal blocks. The approaches used
+    here are of the form 'starting from the input nodes of a DAG, greedily collect
+    the largest block of nodes that match certain criteria'. For additional details,
+    see https://github.com/Qiskit/qiskit-terra/issues/5775.
+    """
+
+    def __init__(self, dag):
+        """
+        Args:
+            dag (Union[DAGCircuit, DAGDependency]): The input DAG.
+
+        Raises:
+            DAGCircuitError: the input object is not a DAG.
+        """
+
+        self.dag = dag
+
+        if isinstance(dag, DAGCircuit):
+            self.is_dag_dependency = False
+
+        elif isinstance(dag, DAGDependency):
+            self.is_dag_dependency = True
```

Comment:
What I've meant is that we have PR #8034, which collects blocks of pairwise commuting gates, meaning that every pair of gates in the same block commutes with each other. This can also be realized with `CollapseChains` interface, however the collection function `collect_function` would be different from the one used to collect blocks of gates matching a given `filter_function`. It's very easy to implement such a function (working on either `DAGCircuit` or `DAGDependency`) -- the block of pairwise-commuting gates is exactly the `front layer` of the DAG, but this particular `collect_function` is not a part of this PR.